### DISCUSSION TITLE
Welcome to Extensions Discussions!  Feedback is highly encouraged.

### DISCUSSION THREAD

**Discussion by cjvandyk** [cjvandyk/Extensions#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

![Visitor Count](https://profile-counter.glitch.me/{cjvandyk}/count.svg)

**Comment by EhsanYzp** [comment]
"Extensions.cs" is an amazing NuGet package using which C# developers can use many "extension methods" for different data types(Array, Dictionary, List, Datetime, Double, Guid, Int16, Int32, Int64, Long, Object, Short, String, Uint16, Uint32, UInt64, ULong, ...)!

There are so many useful features and methods that are available in this package and the best thing about it is that most of these features were implemented as "extension methods" of common data types in the c#/.Net world!

There are also async methods(e.g: for finding Nth prime number) and universal methods(not extensions) for performing different sets of tasks!

The source code is very readable, has a nice and organized structure, and documented well.

Here are some of my thoughts/ideas to improve it:

**Exception Handling:**
- System.Array.cs -> CopyTo: unhandled exceptions (negative values, out of index values)
- Universal.cs -> IsEven/IsOdd/IsPrime/sqrt: unhandled exceptions (negative values) 
- System.Double.cs -> CompoundInterest: unhandled exceptions (negative percentages, percentage more than 100 and negative years)
- System.Object.cs -> Set/Get: unhandled exceptions (e.g: s.Set("", (object)"xyz") , s.Set("s", (object)"xyz"))
- System.String.cs -> Substring: unhandled exceptions (negative and out of index values)
- System.String.cs -> LoremIpsum: if paragraphs > 10, it gives exception. For numbers > 10, you can either set it to 10 (or any other default value) in the method or return!
- System.String.cs -> MorseCodeBeep: if frequency is not in the range, it gives exception. With out of range frequencies, you can set it to min or max permitted frequency(or any other default value) or return!

**Methods Parameters:**
- System.Double.cs -> ToX functions: add precision points as an optional parameter in the methods
- System.String.cs -> HasLower/HasUpper/HasNumberic/HasSymbol: I recommend adding OPTIONAL start and offset parameters (with default values)
- System.String.cs -> IsAlphabetic/IsAlphaNumeric/Ischar/IsLower/IsNumeric/IsUpper/IsLower: I recommend adding OPTIONAL start and offset parameters(with default values)
- System.String.cs -> IsZipCode: I recommend adding country based zip code formats. In the "Constants", you can define an enum for different countries(USA, UK, Canada,...) and then use that enum as one of the method parameters.
- System.String.cs -> Lines: the line seperator can be an OPTIONAL method parameter(right now it is just \n)
- System.String.cs -> DoubleQuote: I recommend adding OPTIONAL offset and length parameters
- System.String.cs -> Words: I recommend adding "seperator char" as an OPTIONAL parameter(with space as default value)

**Methods Return Types:**
- Universal.cs -> ValidateNoNulls: can return bool(false if encountered a null, true otherwise).
-Methods with void return -> I recommend return "bool", with True value upon successful execution; False otherwise.

**Method Definitions:**
- System.String.cs -> Substring: I recommend having two extension methods "SubstringHead" and "SubstringTail" for better readability and usage. Then the last parameter will also not be necessary.
- System.String.cs -> HtmlEncode: I recommend adding another extension invoking "System.Web.HtmlEncode(string s, System.IO.TextWriter output)" method.
- System.String.cs -> HtmlDecode: I recommend adding another extension invoking "System.Web.HtmlDecode(string s, System.IO.TextWriter output)" method.
- System.Guid.cs -> NewCustomGuid: I recommend making this method an extension of either "String" or "Guid" data types.

**Change Non Extension methods to Extensions:**
-Universal.cs -> Bigest: better to be an extension method for different data types(integer, double, ...).
-Universal.cs -> IsEven/IsOdd/IsPrime/sqrt: these are "System.Numerics.BigInteger" extension methods. Better to encapsulate them in a seperated namespace and class.
-Universal.cs -> GetNthPrime/GetNthPrimeAsync: they are not extension methods. I recommend either eliminating them or encapsulate them with a data-type extension method.

**Refactorings:**
- System.Array.cs -> Print: can use the same spacing in the final str(e.g: "  ).
- Universal.cs -> printf: I recommend renaming it to "WritelineWithColor" and wrap the method in another class such as "ConsoleColor"
- Universal.cs -> printf: "if"s are not necessary.
- Universal.cs -> sqrt: missing documentation for sqrt in method definiton.
- Universal.cs -> GetNthPrime/GetNthPrimeAsync: for most of the cases, synchronous version is faster. Better to check the logic behind the code.
- System.DateTime.cs -> ToTimeZone: missing comments for method documentation.
- System.Object.cs: I recommend defining methods working with other datatypes(hashtables, lists, ...) as extensionProperties.

**Other Bugs:**
- Universal.cs -> GetNthPrime/GetNthPrimeAsync: GetNthPrimeAsync(5) & GetNthPrime(5) returns 9 and the correct answer is 11.
- System.Collections.Generic.Dictionary.cs -> ToQueryString: invalid dictionary items(e.g: the ones containing '&','=' or '?' chars) cause building invalid query strings. Better to have some checks/conditions before building the query string.
- System.TimeZoneInfo.cs -> GetTimeZoneString: "time zones" are constant strings in the code. Any modification in the time zones' names should be monitored and reflected to the code.
- System.String.cs -> GetUrlRoot: not working as expected with "http" instead of "https" OR when "http(s)://" part is missing in url
- System.String.cs -> ReplaceTokens: with overlapping key values in dic(1, 11, 111), it is not working as expected
- System.Guid.cs -> NewCustomGuid: not working if the length of string > 8

**General Comments:**
- Add more async extension methods, especially for web-related tasks or the time/resource-consuming ones.
- Add more extension methods for generics.
- Add more web-based extension methods.
- Eliminate all non-extension methods from the package(or if possible, convert some to extension methods). That way, the package will be recognized better.
- Implement "ToX" methods for other types of conversions(distance, length, weight, ...). "ToX" methods were my favorite ones in the package.
- Add unit tests to the project.
- Publish a mini-tutorial project template in Github as a use case of the package.
- Add support for .Net Core/.Net X.
- Separate methods documentation in Github, based on their data type(int, string, DateTime, ...) for better navigation and readability.	
	
I would like to thank you for sharing this amazing project/package with the community. I am sure that it will help lots of C#/.Net developers.

**Reply by cjvandyk** [reply]
A lot of really good feedback @EhsanYzp.  Thank you for that.  I will create issues for bugs you've identified and work to implement recommendations.

**Reply by cjvandyk** [reply]
I had considered exception handling in the extensions when I started, but at the time I was not entirely sure I wanted to handle exceptions in the code instead of letting the exceptions bubble up to the developer.  Having reconsidered the issue, I will implement exception handling and throw clear messages so developers know exactly what went wrong with their invocation of the method.

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
Universal.cs -> IsEven/IsOdd/IsPrime/sqrt: unhandled exceptions (negative values)

Negative value handling added to IsEven(), IsOdd() and IsPrime().  Sqrt() is a Private method that is called from IsPrime(). Since its Private and not exposed in the API and since its caller already validates the parameter value, it would be a waste of CPU cycles to add superfluous value checking to Sqrt().

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
System.Double.cs -> CompoundInterest: unhandled exceptions (negative percentages, percentage more than 100 and negative years)

Percentage value can actually exceed 100% in theory and does cause errors.  Added validation to ensure numbers are non-negative.

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
System.String.cs -> Substring: unhandled exceptions (negative and out of index values)

Added exception handling for negative and out of index values.

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
System.String.cs -> LoremIpsum: if paragraphs > 10, it gives exception. For numbers > 10, you can either set it to 10 (or any other default value) in the method or return!

Added exception handling for negative values and defaulted values larger than 10 to 10.

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
System.String.cs -> MorseCodeBeep: if frequency is not in the range, it gives exception. With out of range frequencies, you can set it to min or max permitted frequency(or any other default value) or return!

Added exception handling for negative values.

**Reply by cjvandyk** [reply]
@EhsanYzp in regard to:
System.Double.cs -> ToX functions: add precision points as an optional parameter in the methods

https://github.com/cjvandyk/Extensions/issues/17
